feat: generateComponents function for React & Solid

OrJDev opened the pull request:
Instead of having to pass the generic everytime, you can now create type safed  components using the `generateReactComponents` method.

```ts
import { generateComponents } from "@uploadthing/react";

import type { OurFileRouter } from "~/server/uploadthing";

export const { UploadButton, UploadDropzone, Uploader } =
  generateComponents<OurFileRouter>();
```

This can later be used like:

```tsx
  <UploadButton
          endpoint="withoutMdwr"
          onClientUploadComplete={(res) => {
            // Do something with the response
            console.log("Files: ", res);
            alert("Upload Completed");
          }}
          onUploadError={(error: Error) => {
            alert(`ERROR! ${error.message}`);
          }}
        />
```

jherr commented:
You should rename `genereateReactComponents` to `generateReactComponents`, the word generate is misspelled.

OrJDev commented:
> You should rename `genereateReactComponents` to `generateReactComponents`, the word generate is misspelled.

Thank you, fixed that and also implemented this function for solid

OrJDev commented:
This also allows the url to be passed at first and then used in all of the solid's adapter functions instead of having to pass it everytime: 

```tsx
import { generateSolidComponents } from "@uploadthing/solid";

import type { OurFileRouter } from "~/server/uploadthing";

const url = "http://localhost:9898";

export const { UploadButton, UploadDropzone, Uploader } =
  generateSolidComponents<OurFileRouter>(url);
```

t3dotgg reviewed (commented):
I think this is worth having in the codebase. cc @markflorkowski @juliusmarminge would love your thoughts

markflorkowski commented:
Yeah, it's nicer than the types components which felt very strange.

jherr commented:
One last thing on naming. If I'm bringing in `createSolidComponents` from a package named `@uploadthing/solid` then the `Solid` is implied. Perhaps `createComponents` in both cases?

```ts
import { generateComponents } from "@uploadthing/react";
```
```ts
import { generateComponents } from "@uploadthing/solid";
```

OrJDev commented:
> One last thing on naming. If I'm bringing in `createSolidComponents` from a package named `@uploadthing/solid` then the `Solid` is implied. Perhaps `createComponents` in both cases?
> 
> 
> 
> ```ts
> 
> import { generateComponents } from "@uploadthing/react";
> 
> ```
> 
> ```ts
> 
> import { generateComponents } from "@uploadthing/solid";
> 
> ```
> 
> 

Yes that is a good idea, will leave that up to @juliusmarminge and @markflorkowski

juliusmarminge commented:
> One last thing on naming. If I'm bringing in `createSolidComponents` from a package named `@uploadthing/solid` then the `Solid` is implied. Perhaps `createComponents` in both cases?
> 
> 
> 
> ```ts
> 
> import { generateComponents } from "@uploadthing/react";
> 
> ```
> 
> ```ts
> 
> import { generateComponents } from "@uploadthing/solid";
> 
> ```
> 
> 

Yes - also matches the api on the backend, we dont do 

```ts

createNextUploadthing from "uploadthing/next"

createNextLegacyUploadthing from "uploadthing/next-legacy"
```

As it's also inferred from the entrypoint's name which runtime you're using.

juliusmarminge commented on .changeset/honest-seas-rush.md:
```diff
@@ -0,0 +1,6 @@
+---
+"@uploadthing/react": patch
+"@uploadthing/solid": patch
```
```suggestion
"@uploadthing/react": minor
"@uploadthing/solid": minor
```

juliusmarminge commented on .changeset/honest-seas-rush.md:
```diff
@@ -0,0 +1,6 @@
+"@uploadthing/solid": patch
+---
+
+feat: generateSolidComponents & generateReactComponents functions
```
```suggestion
feat: `generateComponents` functions for solid and react allows to pass the generic `FileRouter` once instead of for everytime the component is used
```

juliusmarminge commented on docs/src/pages/nextjs/appdir.mdx:
````diff
@@ -67,6 +67,17 @@ export const { GET, POST } = createNextRouteHandler({
 });
 ```
 
+### Creating The UploadThing Components
````
```suggestion
### Creating The UploadThing Components (optional)

Generating components let's you pass your generic `FileRouter` once and then have typesafe components everywhere, instead of having to pass the generic everytime you mount a component, but you can also import the components individually from `@uploadthing/react`.
```

juliusmarminge reviewed (approved):
👍🏼 

left some docs notes

OrJDev commented:
> 👍🏼
> 
> left some docs notes

Resolved everything.
